<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Знакомство-с-данными" data-toc-modified-id="Знакомство-с-данными-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Знакомство с данными</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="возможные-причины-появления-пропусков-в-исходных-данных.-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>возможные причины появления пропусков в исходных данных.</a></span></li></ul></li><li><span><a href="#Шаг-4:-общий-вывод." data-toc-modified-id="Шаг-4:-общий-вывод.-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Шаг 4: общий вывод.</a></span></li></ul></li></ul></div>

# Исследование надежности заемщиков


## Знакомство с данными

In [4]:
import pandas as pd
data = pd.read_csv('datasets/data.csv')

Знакомимся с датасетом

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [6]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [7]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [8]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости вычислим медианное значение трудового стажа `days_employed` в днях.**

In [10]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

In [9]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [11]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [12]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

заполним пропуски медианными значениями

In [13]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [14]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [15]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

In [16]:
data['education'] = data['education'].str.lower()

In [17]:
data.duplicated().sum()

71

In [18]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, добавим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [19]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [20]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [21]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [22]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [23]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

# Исследование данных

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
debt_group=data[data['debt']==1] #выделяем группу должников
non_debt_group=data[data['debt']==0] #выделяем группу тех кто возвращает кредит в срок
debt_total  = debt_group['children'].count() #находим общее количество должников
data_grouped=non_debt_group.groupby(['children']).agg({'debt':'count'}).rename(columns={'debt':'non_debt'})
data_grouped['debt']=debt_group.groupby(['children']).agg({'debt':'count'})
data_grouped['debt']=data_grouped['debt'].fillna(0)
data_grouped['total']=data_grouped['debt']+data_grouped['non_debt']
data_grouped['ratio_debt']=data_grouped['debt']/data_grouped['total']
data_grouped['ratio_non_debt']=data_grouped['non_debt']/data_grouped['total']
data_grouped=data_grouped.sort_values('ratio_debt')
data_grouped

non_debt    debt    total  ratio_debt  ratio_non_debt
children                                                       
5                9     0.0      9.0    0.000000        1.000000
0            13028  1063.0  14091.0    0.075438        0.924562
3              303    27.0    330.0    0.081818        0.918182
1             4364   444.0   4808.0    0.092346        0.907654
2             1858   194.0   2052.0    0.094542        0.905458
4               37     4.0     41.0    0.097561        0.902439

**Вывод:**
Без детей чаще всего укладываются в сроки оплаты по кредитам - 92.4%, после них с тремя детьми - 91.8%, затем с одним ребенком 90.7%, с двумя - 90.5%, с четырьмя - 90.2%. Для клиентов с пятью детьми слишком малая выборка - всего 9 значений.

In [26]:
debt_total  =debt_group['family_status'].count() #находим общее количество должников
data_grouped=non_debt_group.groupby(['family_status']).agg({'debt':'count'}).rename(columns={'debt':'non_debt'})
data_grouped['debt']=debt_group.groupby(['family_status']).agg({'debt':'count'})

data_grouped['total']=data_grouped['debt']+data_grouped['non_debt']
data_grouped['ratio_debt']=data_grouped['debt']/data_grouped['total']
data_grouped['ratio_non_debt']=data_grouped['non_debt']/data_grouped['total']
data_grouped=data_grouped.sort_values('ratio_debt')
data_grouped

non_debt  debt  total  ratio_debt  ratio_non_debt
family_status                                                           
вдовец / вдова              888    63    951    0.066246        0.933754
в разводе                  1105    84   1189    0.070648        0.929352
женат / замужем           11334   927  12261    0.075606        0.924394
гражданский брак           3749   385   4134    0.093130        0.906870
Не женат / не замужем      2523   273   2796    0.097639        0.902361

**Вывод:** "Вдовец/вдова" чаще других укладываются в сроки по кредитам - в 93.3% случаев, за ними идут "в разводе" - 92.9%, "женат/замужем" - 92.4%, "в гражданском браке" - 90.7%. Реже всех укладываются в срок со статусом "Не женат/не замужем" - в 90.2 % случаев.	     

In [29]:
debt_total  =debt_group['total_income_category'].count() #находим общее количество должников
data_grouped=non_debt_group.groupby(['total_income_category']).agg({'debt':'count'}).rename(columns={'debt':'non_debt'})
data_grouped['debt']=debt_group.groupby(['total_income_category']).agg({'debt':'count'})

data_grouped['total']=data_grouped['debt']+data_grouped['non_debt']
data_grouped['ratio_debt']=data_grouped['debt']/data_grouped['total']
data_grouped['ratio_non_debt']=data_grouped['non_debt']/data_grouped['total']
data_grouped=data_grouped.sort_values('ratio_debt')
data_grouped

non_debt  debt  total  ratio_debt  ratio_non_debt
total_income_category                                                   
D                           328    21    349    0.060172        0.939828
B                          4660   354   5014    0.070602        0.929398
A                            23     2     25    0.080000        0.920000
C                         14568  1353  15921    0.084982        0.915018
E                            20     2     22    0.090909        0.909091

#### Как разные цели кредита влияют на его возврат в срок?

In [ ]:
debt_total  =debt_group['purpose_category'].count() #находим общее количество должников
data_grouped=non_debt_group.groupby(['purpose_category']).agg({'debt':'count'}).rename(columns={'debt':'non_debt'})
data_grouped['debt']=debt_group.groupby(['purpose_category']).agg({'debt':'count'})

data_grouped['total']=data_grouped['debt']+data_grouped['non_debt']
data_grouped['ratio_debt']=data_grouped['debt']/data_grouped['total']
data_grouped['ratio_non_debt']=data_grouped['non_debt']/data_grouped['total']
data_grouped=data_grouped.sort_values('ratio_debt')
data_grouped

#### возможные причины появления пропусков в исходных данных.

*Ответ:* Человеческие ошибки при заполнении, переносе, копировании данных. Ошибки при объединении баз данных, в запросах.

### Шаг 4: общий вывод.

Работа проведена. 
Судя по анализу, на вероятность возврата кредита в срок влияет количество детей, семейный статус, размер дохода и цели на которые берется кредит. Но в общем, расхождения всё-таки не кажутся такими уж и значительными в разных категориях.

Лучший заёмщик - бездетный вдова/вдовец с категорией дохода D, который берет его для операций с недвижимостью.
Худший - неженатый/ не замужем с четырьмя детьми, категорией дохода Е, который берет его для операций с автомобилем.

**Дополнение**
Подведём итог ещё раз.
Клиенты, не имеющие детей - более надёжные заёмщики. 
Клиенты, которые официально оформляли свой брак, с большей вероятностью вернут деньги в срок.

Для условности назовём категории дохода A,B,C,D,E следующим образом: высокий, выше среднего, средний, ниже среднего и низкий соответственно.
Клиенты с категорией дохода ниже среднего и выше среднего являются более надежными заемщиками даже по сравнению с категорией среднего дохода. Клиенты с низким доходом и клиенты с высоким доходом реже обращаются за кредитами, что затрудняет сделать однозначные выводы по их категориям.
Кредиты, взятые для операций с недвижимостью и свадеб, с большей вероятностью будут возвращены вовремя, чем те, которые были взяты для образования и автомобилей.